server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java

```diff
@@ -414,10 +427,12 @@ protected void readBytes(ByteBuffer targetBuffer, long position) throws IOExcept
     }
 
     private static void writeCheckpoint(
-            final ChannelFactory channelFactory,
-            final Path translogFile,
-            final Checkpoint checkpoint) throws IOException {
-        Checkpoint.write(channelFactory, translogFile.resolve(Translog.CHECKPOINT_FILE_NAME), checkpoint, StandardOpenOption.WRITE);
+        final FileChannel fileChannel,
+        final Path checkpointFile,
+        final Checkpoint checkpoint) throws IOException {
+        // Since we used a fixed size checkpoint, we overwrite the checkpoint each time.
+        fileChannel.position(0);
```

Comment:
What's a bit subtle here is that we assume callers to not have any concurrent access to this method. Also we assume that once an exception occurs, this method is no more called on the same FileChannel. The surrounding context currently correctly implements this, but it is easy to break and can have disastrous effects. Is there a way we can strengthen these assumptions via assertions?  I was also wondering if we should use positional writes (see `FileChannel.write(ByteBuffer src, long position)`, to ensure we truly always write starting with offset 0 in all cases, and move the positioning logic to the Checkpoint.write method.

Reply:
++ to using positional writes, those should technically be slightly faster as well I think because we only have to lock the internal `positionLock` in the file channel once.